### PR TITLE
feat: popup wallet contact list

### PR DESCRIPTION
**Pull request by hanszhoou**
## Description
https://mask.atlassian.net/browse/MF-4334

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/mask/src/extension/popups/pages/Wallet/Transfer/AddContactInputPanel.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    const { chainId } = useChainContext<NetworkPluginID.PLUGIN_EVM>()
+
+    const theme = useTheme()
+
+    const openAddContactModal = useCallback(() => {
+        AddContactModal.openAndWaitForClose({
+            title: t('wallet_add_contact'),
+            address,
+        })
+    }, [address])
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/Transfer/AddContactInputPanel.tsx]:
```suggestion
    const openAddContactModal = useCallback(async () => {
        return AddContactModal.openAndWaitForClose({
            title: t('wallet_add_contact'),
            address,
        })
    }, [address])
```

File: packages/mask/src/extension/popups/pages/Wallet/Transfer/AddContactInputPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    return (
+        <>
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/Transfer/AddContactInputPanel.tsx]:
```suggestion
```

File: packages/mask/src/extension/popups/modals/AddContactModal/index.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+export type AddContactModalOpenProps = Omit<AddContactModalProps, 'open'>
+export const AddContactModal = forwardRef<SingletonModalRefCreator<AddContactModalOpenProps, boolean>>((_, ref) => {
+    const [props, setProps] = useState<AddContactModalOpenProps>({
+        title: '',
+        address: '',
+    })
+
+    const [open, dispatch] = useSingletonModal(ref, {
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/modals/AddContactModal/index.tsx:129]: Should we use `<ActionModal />` here? CC @UncleBill @nuanyang233

Comment (UncleBill) [reply packages/mask/src/extension/popups/modals/AddContactModal/index.tsx:129]: It can be.

File: packages/shared-base/src/types/Wallet.ts [hunk 4]
```diff
@@ -64,3 +64,8 @@ export interface LegacyWalletRecord {
     createdAt: Date
     updatedAt: Date
 }
+
+export enum WalletContactType {
```
Comment (guanbinrui) [inline packages/shared-base/src/types/Wallet.ts]: It looks the same as `AddressType`.

File: packages/mask/src/extension/popups/modals/AddContactModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    const validationMessage = useMemo(() => {
+        if (addressError) return t('wallets_transfer_error_invalid_address')
+        if (nameAlreadyExist) return t('wallets_transfer_contact_wallet_name_already_exist')
+        return ''
+    }, [t, addressError, nameAlreadyExist])
+
+    const [{ loading }, addContact] = useAsyncFn(async () => {
+        await Web3State.state.AddressBook?.addContact?.(name, address)
+        onConfirm?.()
+    }, [name, address])
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/modals/AddContactModal/index.tsx]:
```suggestion
    }, [name, address, onConfirm])
```

File: packages/mask/src/extension/popups/modals/AddContactModal/index.tsx [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            {validationMessage ? <Typography className={classes.helperText}>{validationMessage}</Typography> : null}
+            <div className={classes.buttonGroup}>
+                <ActionButton className={cx(classes.button, classes.secondaryButton)} onClick={rest.onClose}>
+                    {t('cancel')}
+                </ActionButton>
+                <ActionButton
+                    loading={loading}
+                    onClick={addContact}
+                    className={classes.button}
+                    disabled={addressError || nameAlreadyExist || !name || !name}>
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/modals/AddContactModal/index.tsx]:
```suggestion
                    disabled={addressError || nameAlreadyExist || !name}>
```

File: packages/mask/src/extension/popups/modals/DeleteContactModal/index.tsx [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+function DeleteContactDrawer({ onConfirm, address, name, ...rest }: DeleteContactModalProps) {
+    const { classes, cx } = useStyles()
+    const { t } = useI18N()
+
+    const [{ loading }, deleteContact] = useAsyncFn(async () => {
+        await Web3State.state.AddressBook?.removeContact?.(address)
+        onConfirm?.()
+    }, [address])
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/modals/DeleteContactModal/index.tsx]:
```suggestion
    }, [address, onConfirm])
```

File: packages/mask/src/extension/popups/modals/EditContactModal/index.tsx [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    type: WalletContactType | undefined
+}
+
+function EditContactDrawer({ onConfirm, address, name, setName, type, ...rest }: EditContactModalProps) {
+    const { classes, cx } = useStyles()
+    const { t } = useI18N()
+
+    const contacts = useWalletContacts()
+
+    const nameAlreadyExist = Boolean(contacts?.find((contact) => contact.name === name && contact.address !== address))
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/modals/EditContactModal/index.tsx]:
```suggestion
    const nameAlreadyExist = Boolean(contacts?.find((contact) => contact.name === name && !isSameAddress(contact.address, address)))
```

File: packages/mask/src/extension/popups/modals/EditContactModal/index.tsx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+    const [{ loading }, edit] = useAsyncFn(async () => {
+        if (type === WalletContactType.Contact) {
+            await Web3State.state.AddressBook?.renameContact?.(name, address)
+        } else if (type === WalletContactType.Wallet) {
+            await Web3.renameWallet?.(address, name, { providerType: ProviderType.MaskWallet })
+        }
+
+        onConfirm?.()
+    }, [name, address, type])
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/modals/EditContactModal/index.tsx]:
```suggestion
    }, [name, address, type, onConfirm])
```

File: packages/mask/src/extension/popups/pages/Wallet/ContactList/index.tsx [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+        zIndex: 100,
+    },
+    confirmButton: {
+        margin: '16px 0',
+    },
+}))
+
+const ContactListUI = memo(function TransferUI() {
+    const { t } = useI18N()
+    const { classes } = useStyles({})
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/ContactList/index.tsx:126]:
```suggestion
    const { classes } = useStyles()
```

**Pull request merged**